python/cudf/cudf/core/buffer/buffer.py

```diff
@@ -277,20 +259,147 @@ def memoryview(
+        if offset + size > owner.size:
+            raise ValueError(
+                "offset+size cannot be greater than the size of owner"
+            )
+        self._owner = owner
+        self._offset = offset
+        self._size = size
+
+    @property
+    def size(self) -> int:
+        """Size of the buffer in bytes."""
+        return self._size
+
+    @property
+    def nbytes(self) -> int:
+        """Size of the buffer in bytes."""
+        return self._size
+
+    @property
+    def owner(self) -> Any:
+        """Object owning the memory of the buffer."""
+        return self._owner
+
+    def __getitem__(self, key: slice) -> Self:
+        """Create a new slice of the buffer."""
+        if not isinstance(key, slice):
+            raise TypeError(
+                "Argument 'key' has incorrect type "
+                f"(expected slice, got {key.__class__.__name__})"
+            )
+        start, stop, step = key.indices(self.size)
+        if step != 1:
+            raise ValueError("slice must be C-contiguous")
+        return self.__class__(
+            owner=self._owner, offset=self._offset + start, size=stop - start
+        )
+
+    def get_ptr(self, *, mode: Literal["read", "write"]) -> int:
+        return self._owner.get_ptr(mode=mode) + self._offset
+
+    def memoryview(
+        self, *, offset: int = 0, size: Optional[int] = None
+    ) -> memoryview:
+        size = self._size if size is None else size
+        return self._owner.memoryview(offset=self._offset + offset, size=size)
+
+    def copy(self, deep: bool = True) -> Self:
+        """Return a copy of Buffer.
+
+        Parameters
+        ----------
+        deep : bool, default True
+            - If deep=True, returns a deep copy of the underlying data.
+            - If deep=False, returns a new `Buffer` instance that refers
+              to the same `BufferOwner` as this one. Thus, no device
+              data are being copied.
+
+        Returns
+        -------
+        Buffer
+            A new buffer that either refers to either a new or an existing
+            `BufferOwner` depending on the `deep` argument (see above).
+        """
+
+        # When doing a shallow copy, we just return a new slice
+        if not deep:
+            return self.__class__(
+                owner=self._owner, offset=self._offset, size=self._size
+            )
+
+        # Otherwise, we create a new copy of the memory
+        owner = self._owner._from_device_memory(
+            rmm.DeviceBuffer(
+                ptr=self._owner.get_ptr(mode="read") + self._offset,
+                size=self.size,
+            ),
+            exposed=False,
+        )
+        return self.__class__(owner=owner, offset=0, size=owner.size)
+
+    @property
+    def __cuda_array_interface__(self) -> Mapping:
+        """Implementation of the CUDA Array Interface."""
+        return {
+            "data": (self.get_ptr(mode="write"), False),
```

Comment:
> Will we potentially want the concept of read-only buffer slices

I don't think so

> I don't think it will actually change the mode argument we pass to get_ptr because that's more to do with "do we have control over the lifetime of the usage".

Agree